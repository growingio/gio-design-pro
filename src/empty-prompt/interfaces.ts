import React from 'react';

export interface EmptyPromptProps {
  /**
   自定义 `CSS` 类名
   */
  className?: string;
  /**
   自定义描述内容
   */
  description?: React.ReactNode;
  children?: React.ReactNode;
}
