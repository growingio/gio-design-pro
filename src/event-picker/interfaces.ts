import React from 'react';
import { BasePickerProps } from '../base-picker';
import { BaseProps } from '../utils/interfaces';

export interface Attributes {
  id: string;
  name: string;
  valueType: string;
}

export interface EventData {
  id: string;
  name: string;
  type?: string;
  attributes?: Attributes[];
  itemModels?: Attributes[];
  disabledTips?: React.ReactNode;
  disabled?: boolean;
  selectKey?: string;
  // groupId?: string;
  // groupName?: string;
  [key: string]: unknown;
}
export interface Tab {
  label: string;
  value: string;
}
export interface ListItemPreviewEventProps {
  /**
   * 事件预览中展示图表的回调函数
   */
  onShowEventChart?: (event: EventData) => React.ReactNode;
  /**
   * 事件预览的自定义展示，用于 无埋点事件 详情展示
   */
  previewCustomRender?: (dataSource: EventData) => React.ReactNode;
  /**
   * 获取详情的方法
   */
  fetchDetailData?: (event: EventData) => Promise<EventData>;
  /**
   * 鼠标hover列表项详情面板延迟显示的毫秒数
   */
  detailVisibleDelay?: number;
}
export interface EventPickerProps
  extends Pick<BasePickerProps, 'emptyPrompt' | 'loading'>,
    BaseProps,
    ListItemPreviewEventProps {
  dataSource?: EventData[];
  // /**
  //  * 对数据进行过滤的回调函数
  //  */
  // filter?: (data: EventData[]) => EventData[];
  /**
   * 已选中的项
   */
  value?: EventData | EventData[];
  /**
   * 要显示 Tab 的数据
   */
  tabs?: Tab[];
  /**
   * 是否显示tab['全部']
   */
  showTabAll?: boolean;
  /**
   * 默认的active tab
   */
  activedTab?: string;
  /**
   * 获取对数据进行 tab 级分组的 `key`，
   * 如不提供，则按默认规则取 `key`
   * 最终的值将与 `tabs` 项里的 `value` 进行比较
   */
  getTabKey?: (nodeData: EventData) => string;
  defaultKeyword?: string;
  /**
   * 是否多选
   */
  multiple?: boolean;
  /**
   * 渲染tab footer的回调方法，可以针对不同的tab渲染footer
   */
  panelFooter?: (tabKey: string, data: EventData[]) => React.ReactElement;
  /**
   * 是否隐藏tabNav
   */
  hideTabNav?: boolean;
  /**
   * 本地存储最近使用的属性 key值，用于区分不同用户的最近使用
   */
  historyStoreKey?: string;
  /**
   * 是否重新从localstorage加载最近使用选项,最近使用选项会在点选时自动存储到localstorage,
   * 需要在picker重新渲染或者shouldUpdateRecentlyUsed属性值从false->true时重新加载
   */
  shouldUpdateRecentlyUsed?: boolean;
  /**
   * 鼠标hover列表项详情面板延迟显示的毫秒数
   */
  detailVisibleDelay?: number;
  // /**
  //  * 禁用的选项 id数组
  //  */
  // disabledValues?: string[];
  /**
   * 已选值改变时的回调
   */
  onChange?: (value: EventData | EventData[] | undefined, oldValue?: EventData | EventData[] | undefined) => void;
  /**
   * 选中值时的回调
   */
  onSelect?: (value: EventData | EventData[]) => void;
  /**
   * 列表项点击的回调
   */
  onClick?: (e: EventData) => void;

  /**
   * 多选取消按钮点击的回调
   */
  onCancel?: () => void;
}

export interface EventPickerPreviewProps extends BaseProps, ListItemPreviewEventProps {
  dataSource: EventData;
  // chart?: React.ReactNode;
  // fetchData?: (event: EventData) => Promise<EventData>;
}
