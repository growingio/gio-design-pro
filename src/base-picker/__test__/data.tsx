import React from 'react';
import { Button } from '@gio-design/components';
import { PlusCircleFilled } from '@gio-design/icons';
import List from '../../list';

export const renderItems = () => (
  <>
    <List.Item>内容</List.Item>
    <List.Item>内容</List.Item>
    <List.Item>内容</List.Item>
    <List.Item>内容</List.Item>
    <List.Item>内容</List.Item>
  </>
);

export const renderGroups = () => (
  <>
    <List.ItemGroup title="分组 1">
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
    </List.ItemGroup>
    <List.ItemGroup title="分组 2">
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
    </List.ItemGroup>
    <List.ItemGroup title="分组 3">
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
      <List.Item>内容</List.Item>
    </List.ItemGroup>
  </>
);

export const footer = (
  <Button type="text" icon={<PlusCircleFilled />} size="small">
    新建内容
  </Button>
);

export const tabNavItems = [
  { key: 'tab-1', children: '选项 1' },
  { key: 'tab-2', children: '选项 2' },
];
