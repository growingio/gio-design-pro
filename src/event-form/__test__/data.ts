// eslint-disable-next-line import/prefer-default-export
export const spaceTags: any[] = [
  {
    id: 'JnG40OGz',
    name: '是',
    creator: 'haozhigang',
    creatorId: 'rRGoYQml',
    createdAt: '2021-01-14T07:08:25Z',
    updater: 'haozhigang',
    updaterId: 'rRGoYQml',
    updatedAt: '2021-01-14T07:08:25Z',
    description: '是  得到的',
    platforms: ['minp'],
    docType: 'page',
    actions: ['page'],
    patternMatched: false,
    attrs: {
      content: null,
      index: null,
      href: null,
      path: 'pages/index/index',
      pg: null,
      query: null,
      xpath: null,
      contentType: null,
      domain: 'wx123456',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'wx123456',
      path: 'pages/index/index',
      query: '得到的=s',
      xpath: null,
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: null,
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: { target: '', viewport: '', __typename: 'Screenshot' },
    __typename: 'Element',
  },
  {
    id: 'zZDb6aD9',
    name: '1月13，21次',
    creator: 'lvyuqiang',
    creatorId: 'n1QVaDyR',
    createdAt: '2021-01-13T09:03:46Z',
    updater: 'lvyuqiang',
    updaterId: 'n1QVaDyR',
    updatedAt: '2021-01-13T09:03:46Z',
    description: '1月13，21次',
    platforms: ['minp'],
    docType: 'page',
    actions: ['page'],
    patternMatched: false,
    attrs: {
      content: null,
      index: null,
      href: null,
      path: 'pages/index/index',
      pg: null,
      query: null,
      xpath: null,
      contentType: null,
      domain: 'wx123456',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'wx123456',
      path: 'pages/index/index',
      query: null,
      xpath: null,
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: null,
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: { target: '', viewport: '', __typename: 'Screenshot' },
    __typename: 'Element',
  },
  {
    id: 'mgGJy2pA',
    name: 'wx123 的是页面 pages/form/form 。',
    creator: 'lvyuqiang',
    creatorId: 'n1QVaDyR',
    createdAt: '2021-01-13T07:14:25Z',
    updater: 'lvyuqiang',
    updaterId: 'n1QVaDyR',
    updatedAt: '2021-01-13T07:14:25Z',
    description: '现在定义的是页面 pages/form/form 。',
    platforms: ['minp'],
    docType: 'page',
    actions: ['page'],
    patternMatched: false,
    attrs: {
      content: null,
      index: null,
      href: null,
      path: 'pages/form/form',
      pg: null,
      query: null,
      xpath: null,
      contentType: null,
      domain: 'wx123456',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'wx123456',
      path: 'pages/form/form',
      query: null,
      xpath: null,
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: null,
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: { target: '', viewport: '', __typename: 'Screenshot' },
    __typename: 'Element',
  },
  {
    id: 'AVpZWyGK',
    name: 'ls-xl',
    creator: 'liushuo',
    creatorId: 'qgQMnp3L',
    createdAt: '2021-01-12T05:14:29Z',
    updater: 'liushuo',
    updaterId: 'qgQMnp3L',
    updatedAt: '2021-01-12T05:14:29Z',
    description: '',
    platforms: ['web'],
    docType: 'elem',
    actions: ['chng'],
    patternMatched: false,
    attrs: {
      content: null,
      index: null,
      href: null,
      path: '/push/cdp/testgio-cdpcircel.html',
      pg: null,
      query: null,
      xpath: '/div.main/div.title/form.basic-grey/label/select.selection',
      contentType: null,
      domain: 'release-messages.growingio.cn',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'release-messages.growingio.cn',
      path: '/push/cdp/testgio-cdpcircel.html',
      query: null,
      xpath: '/div.title/form.basic-grey/label/select.selection',
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: null,
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: {
      target: '',
      viewport:
        '/screenshots-uploads/screenshots/07b18645-b784-4abe-b973-a08605bf5893.jpg?sign=wcyjkYl8LHTacSHpdojrXHeTrHPzHONADaFNg5OBY3U%3D&expires=1611037476177',
      __typename: 'Screenshot',
    },
    __typename: 'Element',
  },
  {
    id: 'WkDyn8Qe',
    name: 'ls-jk',
    creator: 'liushuo',
    creatorId: 'qgQMnp3L',
    createdAt: '2021-01-12T05:13:53Z',
    updater: 'liushuo',
    updaterId: 'qgQMnp3L',
    updatedAt: '2021-01-12T05:13:53Z',
    description: '',
    platforms: ['web'],
    docType: 'elem',
    actions: ['clck'],
    patternMatched: false,
    attrs: {
      content: '查找 query 柯基图片',
      index: null,
      href:
        'https://www.google.com/search?q=%E6%9F%AF%E5%9F%BA%E5%9B%BE%E7%89%87&oq=%E6%9F%AF%E5%9F%BA%E5%9B%BE%E7%89%87&aqs=chrome..69i57j0l2j69i61j69i60j0.4959j0j4&sourceid=chrome&ie=UTF-8',
      path: '/push/cdp/testgio-cdpcircel.html',
      pg: null,
      query: null,
      xpath: '/div.main/div.title/form.basic-grey/a',
      contentType: null,
      domain: 'release-messages.growingio.cn',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'release-messages.growingio.cn',
      path: '/push/cdp/testgio-cdpcircel.html',
      query: null,
      xpath: '/div.main/div.title/form.basic-grey/a',
      index: null,
      href: null,
      content: '查找 query 柯基图片',
      pg: null,
      contentType: '=',
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: {
      target: '',
      viewport:
        '/screenshots-uploads/screenshots/364c9f7b-9a27-47c5-94d2-bfe193e3c89a.jpg?sign=bFYchAI2ahfGZykwMOZL3Pm6NBYfc2srQPXu6jt8cC8%3D&expires=1611037476177',
      __typename: 'Screenshot',
    },
    __typename: 'Element',
  },
  {
    id: 'kqQe1qQr',
    name: '定义一个新的事件2',
    creator: 'zhangxueqi',
    creatorId: 'oNGzxQgB',
    createdAt: '2021-01-08T06:21:14Z',
    updater: 'zhangxueqi',
    updaterId: 'oNGzxQgB',
    updatedAt: '2021-01-08T06:21:14Z',
    description: '',
    platforms: ['minp'],
    docType: 'elem',
    actions: ['clck'],
    patternMatched: false,
    attrs: {
      content: '主页',
      index: '0',
      href: 'pages/index/index',
      path: 'pages/index/index',
      pg: null,
      query: null,
      xpath: '#onTabItemTap',
      contentType: null,
      domain: 'wx265d0fa6fa70fae9',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'wx265d0fa6fa70fae9',
      path: 'pages/index/index',
      query: '12=12',
      xpath: '#onTabItemTap',
      index: '0',
      href: 'pages/index/index',
      content: '主页',
      pg: null,
      contentType: '=',
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: { target: '', viewport: '', __typename: 'Screenshot' },
    __typename: 'Element',
  },
  {
    id: '7RDYqYDA',
    name: '事件2的页面',
    creator: 'zhangxueqi',
    creatorId: 'oNGzxQgB',
    createdAt: '2021-01-08T06:20:52Z',
    updater: 'zhangxueqi',
    updaterId: 'oNGzxQgB',
    updatedAt: '2021-01-08T06:20:52Z',
    description: '',
    platforms: ['minp'],
    docType: 'page',
    actions: ['page'],
    patternMatched: false,
    attrs: {
      content: null,
      index: null,
      href: null,
      path: 'pages/index/index',
      pg: null,
      query: null,
      xpath: null,
      contentType: null,
      domain: 'wx265d0fa6fa70fae9',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'wx265d0fa6fa70fae9',
      path: 'pages/index/index',
      query: '12=12',
      xpath: null,
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: null,
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: { target: '', viewport: '', __typename: 'Screenshot' },
    __typename: 'Element',
  },
  {
    id: 'ExQWkgQ7',
    name: '事件4',
    creator: 'zhangxueqi',
    creatorId: 'oNGzxQgB',
    createdAt: '2021-01-08T06:19:40Z',
    updater: 'zhangxueqi',
    updaterId: 'oNGzxQgB',
    updatedAt: '2021-01-08T06:19:40Z',
    description: '',
    platforms: ['minp'],
    docType: 'page',
    actions: ['page'],
    patternMatched: false,
    attrs: {
      content: null,
      index: null,
      href: null,
      path: 'pages/index/index',
      pg: null,
      query: null,
      xpath: null,
      contentType: null,
      domain: 'wx265d0fa6fa70fae9',
      urlScheme: null,
      __typename: 'DocProps',
    },
    definition: {
      domain: 'wx265d0fa6fa70fae9',
      path: 'pages/index/index',
      query: null,
      xpath: null,
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: null,
      urlScheme: null,
      __typename: 'DocProps',
    },
    screenshot: { target: '', viewport: '', __typename: 'Screenshot' },
    __typename: 'Element',
  },
  {
    id: 'xNQXjGMb',
    name: 'autotracker-cdp页面',
    creator: 'mayutong',
    creatorId: 'nxGK0Da6',
    createdAt: '2020-12-15T07:50:07Z',
    updater: 'mayutong',
    updaterId: 'nxGK0Da6',
    updatedAt: '2020-12-15T07:50:07Z',
    description: '',
    platforms: ['Android'],
    docType: 'page',
    actions: ['page'],
    patternMatched: false,
    attrs: {
      content: null,
      index: null,
      href: null,
      path: '/TrackActivity',
      pg: null,
      query: null,
      xpath: null,
      contentType: null,
      domain: 'com.growingio.android.cdp',
      urlScheme: 'growing.447120b8e608d6b9',
      __typename: 'DocProps',
    },
    definition: {
      domain: 'com.growingio.android.cdp',
      path: '/TrackActivity',
      query: null,
      xpath: null,
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: null,
      urlScheme: 'growing.447120b8e608d6b9',
      __typename: 'DocProps',
    },
    screenshot: {
      target: '',
      viewport:
        '/screenshots-uploads/screenshots/2740844b-cb23-4868-8a43-d0f5cd150cc8.jpg?sign=LV5Clhvs36wRzmvx83nGR9J76ZvqDFp%2F1D54C7PixOs%3D&expires=1611037476178',
      __typename: 'Screenshot',
    },
    __typename: 'Element',
  },
  {
    id: 'nxGKXxDa',
    name: '验收限定条件2',
    creator: '马雨彤',
    creatorId: 'evp9bjGO',
    createdAt: '2021-02-08T03:34:14Z',
    updater: '马雨彤',
    updaterId: 'evp9bjGO',
    updatedAt: '2021-02-08T03:34:14Z',
    description: '不勾选',
    platforms: ['Android'],
    docType: 'elem',
    actions: ['clck'],
    patternMatched: false,
    attrs: {
      content: '加载了 并且初始化了 SDK',
      index: null,
      href: null,
      path: '/push/cdp/testgio-cdpcircel.html',
      pg: null,
      query: null,
      xpath: '/div.title/div/div.title/h3',
      contentType: null,
      domain: 'release-messages.growingio.cn',
      urlScheme: 'growing.314b5db864634231',
      __typename: 'DocProps',
    },
    definition: {
      domain: 'release-messages.growingio.cn',
      path: '/push/cdp/testgio-cdpcircel.html',
      query: null,
      xpath: '/div.title/div/div.title/h3',
      index: null,
      href: null,
      content: null,
      pg: null,
      contentType: '=',
      urlScheme: 'growing.314b5db864634231',
      __typename: 'DocProps',
    },
    screenshot: {
      target: '',
      viewport:
        '/screenshots-uploads/screenshots/ecfb4779-46d1-4a10-84b0-1818a9d721d8.jpg?sign=%2FCv2NaAELe7kyLCW7WRrJLckE%2F8X7QL%2FLx5A3FzdG6Y%3D&expires=1612852820607',
      __typename: 'Screenshot',
    },
    __typename: 'Element',
  },
];

export const deviceInfoMinp = { name: '小程序无埋点', domain: 'wx123456', os: 'minp' };
export const currentPageMinp = {
  id: 'f339fcb2bdd07fd7b55020c2835d8bede823e4af',
  eventType: 'PAGE',
  timestamp: 1610954652616,
  ip: '111.198.136.131',
  domain: 'wx123456',
  query: 'name=hello&a=3',
  platform: 'MinP',
  path: 'pages/index/index',
  appType: 0,
  actions: ['clck'],
};
export const deviceInfoApp = {
  projectId: '91eaf9b283361032',
  msgType: 'ready',
  timestamp: 1612755238635,
  domain: 'com.growingio.android.cdp',
  sdkVersion: '3.0.0',
  sdkVersionCode: 30000,
  os: 'Android',
  screenWidth: 1080,
  screenHeight: 1920,
  urlScheme: 'growing.314b5db864634231',
  name: 'Android无埋点测试',
};
export const currentPageApp = {
  path: '/MainActivity',
  title: 'AutoTracker-CDP',
  left: 0,
  top: 0,
  width: 343.125,
  height: 610,
  appType: 1,
  domain: 'com.growingio.android.cdp',
  isDefined: false,
};
