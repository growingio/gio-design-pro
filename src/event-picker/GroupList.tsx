/* eslint-disable react/no-array-index-key */
import { isFunction, keys } from 'lodash';
import React, { useCallback, useMemo, useContext } from 'react';
import { Link } from '@gio-design/components';
import classNames from 'classnames';
import { useLocale, DesignContext } from '@gio-design/utils';
import { EventData, EventPickerProps } from './interfaces';
import List from '../list';
import Group from './Group';
import { GroupItemsProps } from './GroupListItemProps';
import CustomItem from './CustomItem';
import defaultLocale from './locales/zh-CN';
// 类型与名称映射
// export const nameMap: { [key: string]: string } = {
//   history: '最近使用',
//   custom: '埋点事件',
//   simple: '无埋点事件',
//   prepared: '预置计算指标',
//   complex: '自定义计算指标',
//   merged: '合成事件',
// };
export interface Props extends Omit<GroupItemsProps, 'dataSource' | 'keyPrefix'> {
  dataSource: { history?: EventData[]; dataList: Record<string, EventData[]>; select?: EventData[] };
  // value: string[];
  onDeselectAll?: () => void;
  /**
   * 事件预览中展示图表的回调函数
   */
  onShowEventChart?: (event: EventData) => React.ReactNode;
  /**
   * 事件预览的自定义展示，用于 无埋点事件 详情展示
   */
  previewCustomRender?: (dataSource: EventData) => React.ReactNode;
  /**
   * 获取详情的方法
   */
  fetchDetailData?: (event: EventData) => Promise<EventData>;
  /**
   * 是否显示preview 弹出面板
   */
  showPreview?: boolean;
  getGroupName: EventPickerProps['getGroupName'];
  getTypeIcon: EventPickerProps['getTypeIcon'];
  /**
   * 分组的排序方法
   */
  groupSort?: EventPickerProps['groupSort'];
}
const GroupList = (props: Props) => {
  const {
    dataSource: { history = [], dataList, select = [] },
    value = [],
    multiple,
    keyword,
    getGroupName,
    onDeselectAll,
    getTypeIcon,
    groupSort,
    ...rest
  } = props;

  const { locale: { code = 'zh-CN' } = { code: 'zh-CN' } } = useContext(DesignContext);
  const locale = useLocale('EventPicker');
  const { notKnown, rencentUse, clearAllText } = { ...defaultLocale, ...locale } as any;

  const getGroupNameInner = (nodes: EventData[], type: string) => {
    const name = getGroupName?.(type, code) || notKnown;
    return `${name}(${nodes.length})`;
  };
  const handleCheckboxChange = (node: EventData, checked: boolean) => {
    rest.onCheckboxChange?.(node, checked);
  };
  const handleItemClick = (node: EventData) => {
    rest.onClick?.(node);
  };
  const listItems = (items: EventData[], selectValue: string[], keyPrefix = '') =>
    items.map((d: EventData) => {
      const data = (d || {}) as EventData;
      const isSelect = selectValue.includes(data.selectKey || '');
      const listNode = (
        <CustomItem
          {...rest}
          getTypeIcon={getTypeIcon}
          dataSource={data}
          keyword={keyword}
          multiple={multiple}
          value={isSelect ? data.selectKey : undefined}
          onCheckboxChange={(node, checked) => {
            handleCheckboxChange(node, checked);
          }}
          disabled={data.disabled}
          ellipsis
          key={['item', keyPrefix, data.type, data.id].join('-')}
          className={classNames({ selected: isSelect && !multiple })}
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          onClick={(_: any) => handleItemClick(data)}
          detailVisibleDelay={rest.detailVisibleDelay}
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          onMouseEnter={(_) => rest.onMouseEnter?.(data)}
          onMouseLeave={() => rest.onMouseLeave?.()}
          showPreview={rest.showPreview}
          onShowEventChart={rest.onShowEventChart}
        />
      );
      return listNode;
    });
  /**
   * 最近使用分组列表，只展示 没有选中状态
   */
  const renderRecentlyNodes = useMemo(
    () =>
      history?.length > 0 && (
        <React.Fragment key="historyNodes">
          <Group groupKey="history" key="exp-group-history" title={rencentUse}>
            {listItems(history, [], 'history')}
          </Group>
          <List.Divider key="divider-group-history" />
        </React.Fragment>
      ),
    [history]
  );

  const handleClearAll = () => {
    onDeselectAll?.();
  };
  const renderSelectedDataNodes = useCallback(
    () =>
      select?.length > 0 && (
        <React.Fragment key="selectedNodes">
          <Group
            groupKey="selected"
            key="exp-group-selected"
            title={
              <div>
                <Link // eslint-disable-line
                  className="clear-button"
                  component="span"
                  onClick={() => handleClearAll()}
                >
                  {clearAllText}
                </Link>
              </div>
            }
          >
            {listItems(select, value, 'selected')}
          </Group>
          <List.Divider key="divider-group-history" />
        </React.Fragment>
      ),
    [select, value]
  );

  // const groupData = useMemo(() => groupBy(dataList, getTabKey ?? 'type'), [dataList]);
  const orderWeight: { [key: string]: number } = {
    prepared: 90,
    preparedComplex: 80,
    custom: 70,
    simple: 60,
    complex: 50,
  };
  const orderSort = isFunction(groupSort)
    ? groupSort
    : (a: any, b: any) => (orderWeight[b] || 0) - (orderWeight[a] || 0);
  const groupDataNodes = keys(dataList)
    .sort(orderSort)
    .map((key, index) => {
      const items = dataList[key] || [];
      // getListItems(groupData[key]);
      const groupName = getGroupNameInner(items, key);
      return (
        <React.Fragment key={`groupDataNodes-${index}`}>
          {index > 0 && <List.Divider key={`divider-group-${key}-${index}`} />}
          <Group key={`exp-group-${key}`} groupKey={`${key}`} title={groupName} {...rest}>
            {listItems(items, value)}
          </Group>
        </React.Fragment>
      );
    });

  return (
    <List>
      {multiple && renderSelectedDataNodes()}
      {!multiple && renderRecentlyNodes}
      {groupDataNodes}
    </List>
  );
};
export default GroupList;
