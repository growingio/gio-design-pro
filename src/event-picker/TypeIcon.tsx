import React from 'react';

import {
  MetricsPresetOutlined,
  MetricsCustomOutlined,
  CodelessTrackingOutlined,
  EventsPresetOutlined,
  EventsTrackingOutlined,
} from '@gio-design/icons';
import { BaseProps } from '../utils/interfaces';

export type Types =
  | 'custom' // 埋点事件
  | 'simple' // 无埋点事件
  | 'prepared' // 预置计算指标
  | 'complex' // 自定义计算指标
  | 'merged'; // 合成事件

interface Props extends BaseProps {
  type: string;
  size?: string;
}

const TypeIcon: React.FC<Props> = (props) => {
  const { type, ...others } = props;
  switch (type) {
    case 'custom':
    case 'virtual':
      return <EventsTrackingOutlined {...others} />;
    case 'prepared':
      return <EventsPresetOutlined {...others} />;
    case 'simple':
      return <CodelessTrackingOutlined {...others} />;
    case 'preparedComplex':
      return <MetricsPresetOutlined {...others} />;
    case 'complex':
      return <MetricsCustomOutlined {...others} />;
    default:
      return <EventsTrackingOutlined {...others} />;
  }
};

export default TypeIcon;
