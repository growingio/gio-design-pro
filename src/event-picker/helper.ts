import { EventData } from './interfaces';

export { withPrefix } from '@gio-design/components/es/components/cascader/helper';

export const defaultTabs = [
  { label: '事件', value: 'event' },
  { label: '计算指标', value: 'measurement' },
];

// 数据类型映射
export const getEventType = (data: EventData): string => {
  const { type } = data;

  switch (type) {
    case 'simple': // 无埋点事件
    case 'custom': // 埋点事件
    case 'prepared':
    case 'virtual':
    case 'merged':
      return 'event';

    case 'preparedComplex':
    case 'complex':
      return 'measurement';

    default:
      return 'unknow';
  }
};
export const getGroupName = (groupKey: string, locale = 'zh-CN'): string => {
  const nameMap: { [key: string]: string } = {
    custom: '埋点事件',
    virtual: '虚拟事件',
    simple: '无埋点事件',
    preparedComplex: '预置计算指标',
    prepared: '预置事件',
    complex: '自定义计算指标',
    merged: '合成事件',
  };
  const EnglishMap: { [key: string]: string } = {
    custom: 'Custom Events',
    virtual: 'Virtual Events',
    simple: 'Simple Events',
    preparedComplex: 'Prepared Metrics',
    prepared: 'Prepared Events',
    complex: 'Custom Metrics',
    merged: 'Merged Events',
  };
  if (locale === 'en-US') {
    return EnglishMap[groupKey] || 'uknown';
  }
  return nameMap[groupKey] || '未知类型';
};
export const getGroupKey = (data: EventData): string => data.type || 'unkwon';
/**
 * 因为不同类型的事件 id可能重复，此处用 type 和 id 组合出一个唯一key
 * @param data Array<EventData>
 * @returns 包含了 selectKey的新EventData数组
 */
export const withSelectKey = (data: EventData[]) =>
  data.map((d) => {
    const selectKey = `${d.type}-${d.id}`;
    return { ...d, selectKey };
  });
