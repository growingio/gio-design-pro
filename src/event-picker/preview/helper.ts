export const typeMap: { [key: string]: string } = {
  string: '字符串',
  int: '整数',
  float: '小数',
  double: '小数',
};

export const enTypeMap: { [key: string]: string } = {
  string: 'String',
  int: 'Int',
  float: 'Float',
  double: 'Double',
};

export const getTypeName = (type: string) =>
  (localStorage.getItem('locale') === 'en-US' ? enTypeMap[type.toLowerCase()] : typeMap[type.toLowerCase()]) || type;
