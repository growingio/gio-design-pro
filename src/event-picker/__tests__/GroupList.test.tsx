import React from 'react';
import { act } from 'react-dom/test-utils';
import { render, screen, fireEvent } from '@testing-library/react';
import GroupList from '../GroupList';
import { withSelectKey, getGroupName } from '../helper';
import { events } from './data';
import { EventData } from '../interfaces';
import TypeIcon from '../TypeIcon';

jest.useFakeTimers();
const customEvents = withSelectKey((events.filter((e) => e.type === 'custom').slice(0, 15) as unknown) as EventData[]);
describe('<GroupList/> test', () => {
  const onShowEventChart = jest.fn((data) => (
    <div data-testid="gio-chart">
      giochart<div>{data.id}</div>
    </div>
  ));
  const getTypeIcon = (type: string) => <TypeIcon size="14px" className="item-content-icon" type={type || ''} />;

  it('render GroupList', async () => {
    render(
      <GroupList
        getGroupName={getGroupName}
        getTypeIcon={getTypeIcon}
        dataSource={{ dataList: { custom: customEvents } }}
        onShowEventChart={onShowEventChart}
      />
    );

    expect(screen.getByText(/展开全部/)).toBeTruthy();
    act(() => {
      fireEvent.click(screen.getByText(/展开全部/));
    });
    expect(screen.getAllByRole('option')).toHaveLength(customEvents.length);
  });
  it('render multiple select GroupList ', async () => {
    const handleCheckboxChange = jest.fn();
    const handleClick = jest.fn();
    render(
      <GroupList
        multiple
        getGroupName={getGroupName}
        getTypeIcon={getTypeIcon}
        dataSource={{ dataList: { custom: customEvents } }}
        onCheckboxChange={handleCheckboxChange}
        onClick={handleClick}
        onShowEventChart={onShowEventChart}
      />
    );

    act(() => {
      fireEvent.click(screen.getByText(/展开全部/));
    });
    const opts = screen.getAllByRole('option');
    act(() => {
      fireEvent.click(opts[0]);
      fireEvent.click(opts[1]);
    });
    expect(handleClick).toHaveBeenCalledTimes(2);
    expect(handleCheckboxChange).toHaveBeenCalledTimes(2);
  });
  it('trigger mouse event ', async () => {
    const handleItemMouseEnter = jest.fn();
    const handleItemMouseLeave = jest.fn();
    render(
      <GroupList
        getGroupName={getGroupName}
        getTypeIcon={getTypeIcon}
        dataSource={{ dataList: { custom: customEvents } }}
        onMouseEnter={handleItemMouseEnter}
        onMouseLeave={handleItemMouseLeave}
      />
    );

    const opts = screen.getAllByRole('option');
    act(() => {
      fireEvent.mouseEnter(opts[0]);
    });
    act(() => {
      fireEvent.mouseLeave(opts[0]);
    });
    expect(handleItemMouseEnter).toHaveBeenCalledTimes(1);
    expect(handleItemMouseLeave).toHaveBeenCalledTimes(1);
  });

  it('can  deselect all  ', async () => {
    const onDeselectAll = jest.fn();
    const handleCheckboxChange = jest.fn();
    const select = customEvents.slice(0, 2);
    const groupdata = { custom: customEvents.slice(2) };
    render(
      <GroupList
        multiple
        getGroupName={getGroupName}
        getTypeIcon={getTypeIcon}
        dataSource={{ dataList: groupdata, select }}
        onDeselectAll={onDeselectAll}
        onCheckboxChange={handleCheckboxChange}
      />
    );
    act(() => {
      fireEvent.click(screen.getByText('清空全部已选'));
    });

    expect(onDeselectAll).toHaveBeenCalledTimes(1);
  });
});
