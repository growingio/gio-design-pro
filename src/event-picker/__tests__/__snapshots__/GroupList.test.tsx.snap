// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupList/> test render list by custom sort function 1`] = `
NodeList [
  <div
    class="gio-list__item-group__title"
  >
    埋点事件(1)
  </div>,
  <div
    class="gio-list__item-group__title"
  >
    无埋点事件(1)
  </div>,
  <div
    class="gio-list__item-group__title"
  >
    预置事件(1)
  </div>,
]
`;
